core/math/sint/histogram.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (C) 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sint
+
+import (
+	"math"
+	"sort"
+)
+
+type Histogram []int
+
+type HistogramStats struct {
+	Average float64
+	Stddev  float64
+	Median  int
+}
+
+func (h *Histogram) Add(at, count int) {
+	if at < 0 {
+		return
+	}
+	missing := at - len(*h) + 1
+	if missing > 0 {
```

Comment:
You can eliminate the temporary allocation here, and reduce the occurrence of reallocations with something like:

```go

func (h *Histogram) Add(at, count int) {
	if at >= cap(*h) { // at exceeds the slice capacity. Realloc.
		newCap := Max(at*2, 32)
		n := make(Histogram, at+1, newCap)
		copy(n, *h)
		*h = n
	}
	if at >= len(*h) { // at exceeds the slice length. Reslice. 
		*h = (*h)[:at+1]
	}
	(*h)[at] += count
}
```

Reply:
It seems like `append` is already doing exponential reallocations, but yeah cutting down on temporary allocation would be good.